Unite workflow execution api endpoints

guerler opened the pull request:
Resolves #2899, branched from #2935.

guerler committed:
Remove internal api endpoint, merge into invocations endpoint

guerler committed:
Merge branch 'dev' into unite_api

guerler committed:
Convert dict keys to integers

guerler committed:
Fix name generation

guerler committed:
Add details to error modal

guerler committed:
Fix error handling

guerler committed:
Add batch mode test case for invocations api

jmchilton reviewed (changes requested):
I have a couple specific comments also - but at a high-level I was hoping you could throw a flag in there to indicate the response should contain multiple invocations. It seems a bit like magic that if only one invocation if produced a map is returned but if many are produced a list is returned. If the request contained `'multi_request': True` or `batch: True` and then returned a list I would be more comfortable. I'd even more prefer if the batching were done on the client of through the batching stuff Carl wrote - but I understand the pragmatic concerns here.

guerler commented:
Feel free to correct but if I understand this right, this is already the case. A list of invocations is only returned if one of the parameters has specified `batch: True` otherwise a single map is returned. Besides that, can you help with the test case, it works locally with sqllite and postgres but somehow fails when ran through jenkins. Any ideas? Thanks in advance.

guerler committed:
Fix phrasing of error message

guerler committed:
Revise target url

guerler committed:
Always return a list for formats with batch mode option

guerler committed:
Improve invocations success message

guerler committed:
Remove unused history wait

guerler committed:
Use hid from hda

guerler committed:
Fix indent

guerler committed:
Merge all errors from batch mode run for regular tool executions

guerler committed:
Use order_index instead of step_id

guerler commented:
The request now always returns a list if `input_params` is specified, https://github.com/galaxyproject/galaxy/pull/2969/commits/1cb67308771edf40716ef5f30f4321051debc4bb#diff-2fca1f4002bba50737993fd85ce5800fR454.

jmchilton commented:
@guerler I'd really like to unify the `input_params` with the existing `parameters` approach. I don't want the batch mode options to be available in one and not the other - it also isn't very intuitive from the parameter name `input_params` that if it is specified you get a list back. I know it feels redundant in some ways, but I think it would be better if the UI explicitly asked for a list back (`enable_batch_processing: True` or `batch: True`). It would be easier for the endpoint consumer to understand and it would enable us to make more progress toward uniting `input_params` and `parameters`.

guerler committed:
Use parameters/batch instead of input_params

guerler committed:
Rename test associated workflow export

nsoranzo commented:
It would be nice to test if this breaks BioBlend before merging.

jmchilton reviewed (commented):
Sorry if I seem picky - I do appreciate this particular change - I think it is important improvement.

jmchilton committed:
Increased consistency in backend workflow run code.

Rename build_workflow_run_config to build_workflow_run_configs since it sometimes returned a list, make it always return a list, one path out of run_request.py:build_workflow_run_configs, ensure batch parameter is respected and throw an exception if it is not and build_workflow_run_configs returns multiple invocations.

jmchilton commented:
@guerler - about this...

```
new_history.copy_tags_from( trans.user, trans.history )
```

is that intentional? That seems wrong - but was it copied from the old code in the UI controller or something?

guerler committed:
Do not copy tags from source history

guerler commented:
@jmchilton its not a big deal.

jmchilton commented:
@guerler 

```
                if len( param_keys ) > index:
                    ids = param_keys[ index ]
```

This should just be:

```
                if len( param_keys ) > 0:
                    ids = param_keys[ index ]
```

Or something like that right? param_keys is the hids of things being batched up right?

jmchilton committed:
Unify workflow API history options across approaches.

The old variant of specifying a history attribute should work with new batch parameter mode and the new variant of specifing new_history_name or history_id should work with the existing invocation API stuff.

guerler committed:
Add error response to validate expansion result

jmchilton committed:
Merge remote-tracking branch 'guerler/unite_api' into unite_api_history

Conflicts:
	lib/galaxy/workflow/run_request.py

jmchilton commented:
I opened some PRs to merge approaches to history handling and make the return value of the run_request method more consistent. Let me know if either is objectionable. I haven't tested them extensively yet but I'd like to know if you are okay with general directions of the refactorings.

guerler commented:
Alright, I'll take a look. The conditional double checks the result of the parameter expansion. We could remove it entirely, but I do not see any harm leaving it in. The conditional should never fail unless there is a bug in the parameter expansion procedure.

jmchilton commented:
@guerler I see - that makes sense now - I was mistaking this as the `if nids == 0` case - but  I see that is just taken care of but not doing anything in that nested conditional. I'd drop the new exception and place an assertion in there instead of a conditional - it feels like the most Pythonic thing to do.

Update: This is what I would do (https://github.com/guerler/galaxy/pull/13/commits/ab2a5198afeb0b651643b84c0dd3eb6a0e611bb6) I guess.

jmchilton committed:
Resolve this comment/discussion (https://github.com/galaxyproject/galaxy/pull/2969#issuecomment-250830186).

jmchilton committed:
Revert to an exception instead of an assertion.

I'm close to winning @guerler over - but not quite there yet.

jmchilton committed:
Merge remote-tracking branch 'guerler/unite_api' into unite_api_history

Conflicts:
	lib/galaxy/workflow/run_request.py

jmchilton commented:
@nsoranzo - I'll definitely run the BioBlend tests and I'll PR enhancements coming out of this work -descriptions of the batch functionality and the flattened parameter style @guerler prefers here.

jmchilton committed:
Fixup previous commits bugs.

guerler committed:
Use explicit order_index when assigning workflow steps

guerler committed:
Test case

jmchilton commented:
@guerler - You can sort keys before iterating over them during expansion to make it deterministic right?

jmchilton commented:
@guerler I just noticed you are exposing unencoded IDs with the new run export. Can we fix that before the release? I feel strongly that we shouldn't be exposing unencoded IDs over the API - it has long been a design goal of the API I think. It should be possible to represent everything with `order_index` - to whatever extent we can use that the API is more usable and stable to client libraries and humans hacking around.

jmchilton committed:
Merge remote-tracking branch 'guerler/unite_api' into unite_api_history

jmchilton committed:
Refactor run_request to reuse step parameter normalization.

Lessens the difference between GUI and API approaches.

jmchilton committed:
Refactor build_workflow_run_configs toward unification of API and UI approaches.

jmchilton committed:
Refactor build_workflow_run_configs toward unification of API and UI approaches.

Swap expansion of raw parameters and normalization of parmeters so that we can move the loop outside of the conditional ... and thus elimiante the conditional that distinguishes between API and UI approaches.

jmchilton committed:
Refine workflow target history handling with bug fix and improved error message.

jmchilton committed:
Unify build_workflow_run_configs for API and UI approaches to running workflows.

No more branches - the method runs top to bottom for each approach. One loop for batching regardless and completely available to all API options. Only call WorkflowRunConfig in one place. Remove duplicate fetching of replacement_params. One call to _normalize_step_parameters.

jmchilton commented:
https://github.com/guerler/galaxy/pull/13 I think includes a bunch of refactoring so that instead of branching based on approach in `build_workflow_run_config` at the very outer part of a huge function and doing very different things and handling different parameters - everything is handled in one big loop with more consistent parameter handling. The following parameters `history`, `history_id`, `history_id`, `inputs`, `ds_map`, `inputs_by`, `legacy`,  and `no_add_to_history` are now handled consistently across approaches (UI, existing API) - before https://github.com/guerler/galaxy/pull/13 each of these parameters would only apply to one branch or the other of this code. In addition initial `WorkflowRunConfig` objects to be persisted are now only created in one place - one line of code - I think this is strong evidence that we now just have that one path through the code.

The thing that this part of the run UI continues to do which I think is unfortunate is write IDs to JSON blobs in `WorkflowRequestStepState.state` for input steps instead of properly using `WorkflowRequestToInputDatasetAssociation` or `WorkflowRequestToInputDatasetCollectionAssociation` model objects / tables which provide referential integrity and stronger typing since these use the database more appropriately. I don't think that issue needs to hold this up though.

guerler committed:
Merge pull request #13 from jmchilton/unite_api_history

Unify workflow API history options across approaches.

guerler committed:
Encode unencoded step id when exposing workflow dictionary for execution

guerler committed:
Move get target history out of build_workflow_run_configs

guerler committed:
Refactor target history mode validation, remove spaces

jmchilton commented on lib/galaxy/managers/workflows.py:
```diff
@@ -388,12 +388,12 @@ def _workflow_to_dict_run( self, trans, stored ):
                     'inputs' : [ input.to_dict( trans ) for input in inputs.itervalues() ]
                 }
-            step_model[ 'step_id' ] = step.id
+            step_model[ 'step_id' ] = trans.app.security.encode_id( step.id )
```
Do you have a reason not to just use the `order_index` for these? The other export endpoints do this and I think we should be consistent - I also think this makes the API more "usable". I can't think of a reason the GUI should need to know the step ID when the invocation API doesn't use it and the workflow editor doesn't use it - is there one?

guerler replied:
This id is used to identify connected steps. I used the id because its native to the WorkflowStepConnection model (used below). However, anything which uniquely identifies the steps/connected steps would work.

guerler committed:
Use order index to link connected datasets

guerler committed:
Rename step_id to step_index to avoid confusion

guerler committed:
Rename step_id to step_index to avoid confusion, fix error message response handling

guerler committed:
Improve error display robustness for run workflow form

guerler committed:
Remove sorting from workflow expansion test cases

guerler committed:
Remove sorting from last expansion test case

guerler committed:
Fix order in workflow expansion

guerler committed:
Rely on deterministic invocations order in batch mode test case

guerler commented:
@galaxybot test this

guerler commented:
@galaxybot test this

jmchilton commented:
You disappeared from IRC:

```
01:20 <jmchilton> I think you want to sort the keys - not the items 
01:21 <jmchilton> maybe it doesn't matter - but I don't know what sorting a tuple like 
                  that does 
01:22 <jmchilton> I mean - I guess it work on the first item... hmmm... 
01:25 <jmchilton> Okay - I guess the thing is that the invocations do get queued -  
                  there is no guarentee which will start adding to the history first 
01:25 <jmchilton> it is like submitting jobs in a given order and expecting them to  
                  run in the same order - Galaxy doesn't do that 
01:26 <jmchilton> I guess I don't consider that a problem - we could work on FIFO-ing  
                  the workflow scheduler though 
```

guerler commented:
Alright. Just wanted to make sure that testing both invocation orders is really the right thing to do. Thx.

guerler committed:
Check both invocation orders when using the new batch mode feature

martenson commented:
I tried to do some black box testing here and when I try to run a simple wf on this branch it instantly kills my browser tab.

![screenshot 2016-10-04 11 05 52](https://cloud.githubusercontent.com/assets/1814954/19079740/a0766ae4-8a22-11e6-8d7e-2c40fac2b346.png)

![screenshot 2016-10-04 11 06 12](https://cloud.githubusercontent.com/assets/1814954/19079746/a854558c-8a22-11e6-8d65-ed276ff5f8f9.png)

![screenshot 2016-10-04 11 05 28](https://cloud.githubusercontent.com/assets/1814954/19079754/ad45bb12-8a22-11e6-86ef-72148a727b97.png)

martenson commented:
update: Chrome with open devtools causes the issue above, with devtools closed it runs fine

Another issue is that if you open the run workflow in incognito window it will fail to render properly

![screenshot 2016-10-04 12 23 29](https://cloud.githubusercontent.com/assets/1814954/19083005/6cca4a9c-8a2e-11e6-8298-457b0e2e2de1.png)

guerler committed:
Hide children elements instead of destroying them upon successful workflow submission

guerler commented:
Very good catch. Hiding the child forms instead of removing them avoids the crash when Chromes devtools are open. I could not reproduce the second issue yet.

martenson commented:
I can produce the incognito issue reliably on a WF that has only the 'merge columns' tool in it.

martenson commented:
as a sideeffect of 329ee052510c0c720ae5d44bac23becf639f6506 the submission is _way_ faster

martenson commented:
the second issue was caused by a broken WF, when exported it has no steps:

``` json
{
    "a_galaxy_workflow": "true", 
    "annotation": "", 
    "format-version": "0.1", 
    "name": "merge test", 
    "steps": {}, 
    "uuid": "87df19a4-454d-445b-ba2b-0f35676e14a0"
}
```

I think we should make WF creation atomic operation, but that is not in scope of this PR.

martenson reviewed (approved):
Looks and works very fine to me. Thank you duo of  @guerler & @jmchilton

jmchilton commented:
This branch seems to cause a bioblend test failure (most though not 100% of the time).

Against dev:

```
BIOBLEND_GALAXY_API_KEY=a7bf49fe4fd4041a2a93de4d0901716b BIOBLEND_GALAXY_URL=http://localhost:8080/ nosetests tests/TestGalaxyWorkflows.py
test_cancelling_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_export_workflow_to_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_get_workflows (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_import_workflow_from_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_run_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_show_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok

----------------------------------------------------------------------
Ran 7 tests in 15.312s

OK
```

Against this branch:

```
BIOBLEND_GALAXY_API_KEY=a7bf49fe4fd4041a2a93de4d0901716b BIOBLEND_GALAXY_URL=http://localhost:8080/ nosetests tests/TestGalaxyWorkflows.py
test_cancelling_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_export_workflow_to_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_get_workflows (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_import_workflow_from_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_run_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_show_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... FAIL

======================================================================
FAIL: test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/john/workspace/bioblend/tests/test_util.py", line 54, in wrapped_method
    return method(has_gi, *args, **kwargs)
  File "/home/john/workspace/bioblend/tests/test_util.py", line 54, in wrapped_method
    return method(has_gi, *args, **kwargs)
  File "/home/john/workspace/bioblend/tests/TestGalaxyWorkflows.py", line 54, in test_workflow_scheduling
    self.assertEqual(invocation['state'], "ready")
AssertionError: u'new' != 'ready'
-------------------- >> begin captured logging << --------------------
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/tools?in_panel=False&key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/tools?in_panel=False&key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/workflows/upload?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/histories?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/tools?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/workflows/59c76a119581e190/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 7 tests in 25.187s

FAILED (failures=1)
```

After dropping that assertion and letting the test continue:

```
BIOBLEND_GALAXY_API_KEY=a7bf49fe4fd4041a2a93de4d0901716b BIOBLEND_GALAXY_URL=http://localhost:8080/ nosetests tests/TestGalaxyWorkflows.py
test_cancelling_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_export_workflow_to_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_get_workflows (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_import_workflow_from_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_run_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_show_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ERROR

======================================================================
ERROR: test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/john/workspace/bioblend/tests/test_util.py", line 54, in wrapped_method
    return method(has_gi, *args, **kwargs)
  File "/home/john/workspace/bioblend/tests/test_util.py", line 54, in wrapped_method
    return method(has_gi, *args, **kwargs)
  File "/home/john/workspace/bioblend/tests/TestGalaxyWorkflows.py", line 57, in test_workflow_scheduling
    pause_step = steps[2]
KeyError: 2
-------------------- >> begin captured logging << --------------------
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/tools?in_panel=False&key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/tools?in_panel=False&key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/workflows/upload?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/histories?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/tools?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/workflows/5a9381f02a7b89af/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 7 tests in 22.767s

FAILED (errors=1)
```

The failing test https://github.com/galaxyproject/bioblend/blob/master/tests/TestGalaxyWorkflows.py#L25. I wrote the test and am quite surprised it is testing something not tested by Galaxy itself - I'll look into this some more.

jmchilton commented:
Appears to be something up with the history id being empty on the resulting invocation.

jmchilton committed:
Fix for unite_api branch for API key specification without explicit history target.

It is good to not depend on trans.history being set I think and this is required to fix bioblend tests.

guerler committed:
Merge pull request #14 from jmchilton/unite_api_apikey_fix

Fix for unite_api branch for API key specification without explicit history target.

guerler committed:
Always submit either history_id or new history name through run form

guerler committed:
Remove trans.history from target history identification
